### DISCUSSION TITLE
Memory Layout for VariableSizedData - Strings

### DISCUSSION THREAD

**Discussion by pfeiffer-tim** [nebulastream/nebulastream#528] [Ideas]
### General:
___
Storing and accessing variable-sized data requires a dedicated memory layout which is designed based on specifications for _data characteristics_, _performance_, _storage efficiency_, and _query access patterns_. 

Types of variable-sized data we're interested in: 
- Strings: Textual data of varying length, e.g. text fields (VARCHAR) or document fields (JSON stored as strings).
- Binary Data (BLOB): Arbitrary binary content with varying sizes, e.g. in image/videos/audio files.
- Nested or Hierarchical Data: Data with a flexible structure, often represented as trees or graphs, e.g. JSON/ARRAY/LIST.

In addition, other types such as _Geospatial Data_ and _Time-Series Data_ exist, but we should focus on **Strings** and **BLOBs** first, as these two types can cover many use cases for NebulaStream.
### Discussion:
___
The discussion here is related to **Variable-Length String Data**. More precisely, we want to understand the following questions w.r.t to the **data characteristics**:
1. How does string data actually looks like in the IoT environment?
2. What kind of string data are we particularly interested in? _Long vs. short strings, repeated vs. unique strings, etc._  

If we understand the basic characteristics on textual data for NebulaStream, we can work backwards from here and optimize based on the related requirements.

### Memory Layouts:
___
The NebulaStream team has already started an offline discussion on two specific memory layouts that are of particular interest for string and binary data. The following reviews both approaches with simple figures for visualization:

### Layout 1 aka "BLOB layout":
_![BLOB_layout](https://github.com/user-attachments/assets/23e1ecec-3b06-4676-860b-fb2c65449ad4)_
**Approach:**
Inside the tuple buffer, a string data type field consists of its size (4 bytes) and a pointer (8 bytes) to the actual data. The pointer references to the beginning of its character sequence which is stored in another tupleBuffer (here called string heap). This layout can also be used for binary data.

**Pros:**
- Can be used for large+short strings and binary data.
- Enables efficient compression as we store a big chunk of binary data (batch processing during encoding/decoding).

**Cons:**
- Requires its own memory management strategy for the additional allocated string heaps.
- Expensive "childBuffer" pointer dereference operations when accessing the actual data.
- Sending the actual data (here string heaps) to other nodes in a distributed environment could be more complex w.r.t the implementation.

### Layout 2 aka "Short-String layout":
![short_string_layout](https://github.com/user-attachments/assets/3f224b85-03b9-4492-ae79-785a6d271c24)
**Approach:**
Inside the tuple buffer, a string data type field consists only of its size (4 bytes). The actual data is stored at the end of the tupleBuffer and grows towards the records. This means we have 2 pointers, one for the records, one for the string data. When both pointers cross, we know that we need a new tupleBuffer to store the record. If we want to access the actual string data of a record, we can use the record offset/index and then its string length to know how far we need to read.

**Pros:**
- All the data is stored inside one single TupleBuffer, no "childBuffer-chaining" needed, i.e. no extra memory management.
- No additional pointer dereference operations.
- Suitable for short strings.

**Cons:**
- Potentially less room for efficient compression if the tupleBuffer consists of only small number of strings.
- If we process large binary data, or large strings: One data item could span across multiple tupleBuffers. We end up with multiple tupleBuffers that do not hold any other records, just the overflowed data.

### Goal:
___
In the future, it would be nice to support multiple memory layouts for different variable-sized data types w.r.t. their data characteristics, e.g. one for short strings, another one for BLOBs. Ideally, there should be a simple interface like _readVarSizedValue()_ / _writeVarSizedValue()_ that hides the underlying format so that caller does not need to worry about how to store & access values efficiently.

**Comment by pfeiffer-tim** [comment]
Again, let's first focus on the questions related to string data characteristics:
- How does string data actually looks like in the IoT environment?
- What kind of string data are we particularly interested in? Long vs. short strings, repeated vs. unique strings, etc.

**Comment by keyseven123** [comment]
Couple of questions: 
- You call the two proposed variants `Short String Layout` and `BLOB Layout`. I assume that the `BLOB Layout` is also the one for longer strings?
- Can the different version be changed on the fly or does this require recompilation? Or in another way, do we change the layout for one query/pipeline or for one tuple buffer? How do we know this?

**Comment by yschroeder97** [comment]
I have the feeling that this model with two directions introduces a lot of edge cases and difficult scenarios, whereas the first one seems more straightforward and in line with what something like DuckDB uses. 
With the second model, you need to have a mechanism to handle large strings anyway, since it could happen really fast that only a handful of tuples fit into one buffer when their variable length-data is large.
If you opt for the first solution, you can later add the Umbra string layout potentially as an optimization on top.

Also, what layouts do the following systems use:
- RisingWave
- Flink
- Spark Streaming
- ClickHouse
 
I have a feeling that there are more options than these two. 

Although in general, string data seems to be very important as users tend to dump their data into strings or JSON documents, I am not sure if this holds for IoT data.
